src/batch.rs

```diff
@@ -27,61 +24,130 @@ fn gen_128_bits<R: RngCore + CryptoRng>(mut rng: R) -> [u64; 4] {
+    inner: Inner,
 }
 
-impl<'msg, M: AsRef<[u8]>, T: SigType + ?Sized>
-    From<(VerificationKeyBytes<T>, Signature<T>, &'msg M)> for Item<T>
+impl<'msg, M: AsRef<[u8]>>
+    From<(
+        VerificationKeyBytes<SpendAuth>,
+        Signature<SpendAuth>,
+        &'msg M,
+    )> for Item
 {
-    fn from(tup: (VerificationKeyBytes<T>, Signature<T>, &'msg M)) -> Self {
+    fn from(
+        tup: (
+            VerificationKeyBytes<SpendAuth>,
+            Signature<SpendAuth>,
+            &'msg M,
+        ),
+    ) -> Self {
         let (vk_bytes, sig, msg) = tup;
         // Compute c now to avoid dependency on the msg lifetime.
         let c = HStar::default()
             .update(&sig.r_bytes[..])
             .update(&vk_bytes.bytes[..])
             .update(msg)
             .finalize();
-        Self { vk_bytes, sig, c }
+        Self {
+            inner: Inner::SpendAuth { vk_bytes, sig, c },
+        }
     }
 }
 
+impl<'msg, M: AsRef<[u8]>> From<(VerificationKeyBytes<Binding>, Signature<Binding>, &'msg M)>
+    for Item
+{
+    fn from(tup: (VerificationKeyBytes<Binding>, Signature<Binding>, &'msg M)) -> Self {
+        let (vk_bytes, sig, msg) = tup;
+        // Compute c now to avoid dependency on the msg lifetime.
+        let c = HStar::default()
+            .update(&sig.r_bytes[..])
+            .update(&vk_bytes.bytes[..])
+            .update(msg)
+            .finalize();
+        Self {
+            inner: Inner::Binding { vk_bytes, sig, c },
+        }
+    }
+}
+
+// This would ideally be an associated type with `Verifier` but
+// generic associated types are unstable:
+// https://github.com/rust-lang/rust/issues/44265
+type BatchTuple<T: SigType> = (VerificationKeyBytes<T>, Signature<T>, Scalar);
+
 /// A batch verification context.
-pub struct Verifier<T: SigType> {
+pub struct Verifier {
     /// Signature data queued for verification.
-    signatures: Vec<(VerificationKeyBytes<T>, Scalar, Signature<T>)>,
+    signatures: Vec<Item>,
     /// Caching this count avoids a hash traversal to figure out
     /// how much to preallocate.
     batch_size: usize,
 }
 
-impl<T: SigType> Default for Verifier<T> {
-    fn default() -> Verifier<T> {
+impl Default for Verifier {
+    fn default() -> Verifier {
         Verifier {
             signatures: vec![],
             batch_size: usize::default(),
         }
     }
 }
 
-impl<T: SigType> Verifier<T> {
+impl Verifier {
     /// Construct a new batch verifier.
-    pub fn new() -> Verifier<T> {
+    pub fn new() -> Verifier {
         Verifier::default()
     }
 
-    /// Queue a (key, signature, message) tuple for verification.
-    pub fn queue<I: Into<Item<T>>>(&mut self, item: I) {
-        let Item { vk_bytes, sig, c } = item.into();
-
-        self.signatures.push((vk_bytes, c, sig));
+    /// Queue an Item for verification.
+    pub fn queue<I: Into<Item>>(&mut self, item: I) {
+        self.signatures.push(item.into());
         self.batch_size += 1;
```

Comment:
just double checking, should this always be equal to `self.signatures.len()` and if so can we remove `self.batch_size` and replace it with `fn batch_size(&self) -> usize { self.signatures.len() }`?

Reply:
I think this is true for the latest changes, looks like your looking at the outdated ones?

Reply:
:thumbsup:, for some reason I can't mark things as resolved but yea I see it now

Reply:
cf https://github.com/ZcashFoundation/redjubjub/pull/36#discussion_r449676584 ; there's no need for a `fn batch_size` at all, I think.